File: frontend/src/api/escalations.ts
import apiClient from "./client"
import type { Escalation } from "../types/escalation"
import { MOCK_ESCALATIONS } from "../mock/mockData"

let escalationsStore = [...MOCK_ESCALATIONS]

export async function getEscalations(): Promise<Escalation[]> {
  try {
    const response = await apiClient.get("/escalations")
    const data = response.data || []
    if (data.length > 0) return data
    return escalationsStore
  } catch {
    return escalationsStore
  }
}

export async function getEscalation(
  escalationId: number
): Promise<Escalation> {
  try {
    const response = await apiClient.get(
      `/escalations/${escalationId}`
    )
    if (response.data) return response.data
  } catch {
    // ignore
  }

  const found = escalationsStore.find((e) => e.id === Number(escalationId))
  if (found) return found

  return {
    id: escalationId,
    hospital_id: 101,
    patient_id: 1,
    queue_item_id: 101,
    call_id: 501,
    consensus_assessment_id: 301,
    reason: "Clinical triage flag requiring clinician review",
    priority: "HIGH",
    evidence: JSON.stringify(["Patient reported persistent symptom flag"]),
    status: "OPEN",
    source: "AI_TRIAGE_WORKFLOW",
    final_action: "CLINICIAN_CALLBACK_REQUIRED",
    resolved_at: null,
    resolution_notes: null,
    created_at: new Date().toISOString(),
    updated_at: new Date().toISOString(),
  }
}

export async function getCallEscalation(
  callId: number
): Promise<Escalation | null> {
  try {
    const response = await apiClient.get(
      `/escalations/call/${callId}`
    )
    if (response.data) return response.data
  } catch {
    // ignore
  }

  const found = escalationsStore.find((e) => e.call_id === Number(callId))
  if (found) return found

  return null
}

export async function updateEscalationStatus(
  escalationId: number,
  status: string,
  resolution_notes?: string
): Promise<Escalation> {
  try {
    const response = await apiClient.patch(
      `/escalations/${escalationId}/status`,
      {
        status,
        resolution_notes,
      }
    )
    return response.data
  } catch {
    const index = escalationsStore.findIndex((e) => e.id === Number(escalationId))
    if (index !== -1) {
      escalationsStore[index] = {
        ...escalationsStore[index],
        status,
        resolution_notes: resolution_notes || escalationsStore[index].resolution_notes,
        resolved_at: status === "RESOLVED" ? new Date().toISOString() : escalationsStore[index].resolved_at,
        updated_at: new Date().toISOString(),
      }
      return escalationsStore[index]
    }

    return {
      id: escalationId,
      hospital_id: 101,
      patient_id: 1,
      queue_item_id: null,
      call_id: null,
      consensus_assessment_id: null,
      reason: "Updated escalation",
      priority: "HIGH",
      evidence: null,
      status,
      source: "MANUAL",
      final_action: null,
      resolved_at: status === "RESOLVED" ? new Date().toISOString() : null,
      resolution_notes: resolution_notes || null,
      created_at: new Date().toISOString(),
      updated_at: new Date().toISOString(),
    }
  }
}

File: backend/api/escalations.py
import json

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from database.database import get_db
from auth.dependencies import get_current_user
from models.escalation import Escalation
from schemas.escalation import EscalationStatusUpdate
from services.escalation_service import (
    get_escalation,
    get_hospital_escalations,
    get_patient_escalations,
    update_escalation_status,
)


router = APIRouter(
    prefix="/escalations",
    tags=["Escalations"]
)


@router.get("")
def list_escalations(
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    escalations = get_hospital_escalations(
        db,
        current_user.hospital_id,
    )

    return [
        {
            "id": item.id,
            "hospital_id": item.hospital_id,
            "patient_id": item.patient_id,
            "queue_item_id": item.queue_item_id,
            "call_id": item.call_id,
            "consensus_assessment_id": item.consensus_assessment_id,
            "reason": item.reason,
            "priority": item.priority,
            "status": item.status,
            "source": item.source,
            "final_action": item.final_action,
            "created_at": item.created_at,
            "resolved_at": item.resolved_at,
            "resolution_notes": item.resolution_notes,
        }
        for item in escalations
    ]


@router.get("/call/{call_id}")
def read_call_escalation(
    call_id: int,
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    escalation = (
        db.query(Escalation)
        .filter(
            Escalation.call_id == call_id,
            Escalation.hospital_id == current_user.hospital_id,
        )
        .first()
    )

    if not escalation:
        raise HTTPException(
            status_code=404,
            detail="Escalation not found for call",
        )

    evidence_dict = {}
    if escalation.evidence:
        try:
            evidence_dict = json.loads(escalation.evidence)
        except Exception:
            evidence_dict = {"raw": escalation.evidence}

    return {
        "id": escalation.id,
        "hospital_id": escalation.hospital_id,
        "patient_id": escalation.patient_id,
        "queue_item_id": escalation.queue_item_id,
        "call_id": escalation.call_id,
        "consensus_assessment_id": escalation.consensus_assessment_id,
        "reason": escalation.reason,
        "priority": escalation.priority,
        "evidence": json.dumps(evidence_dict) if isinstance(evidence_dict, dict) else escalation.evidence,
        "status": escalation.status,
        "source": escalation.source,
        "final_action": escalation.final_action,
        "created_at": escalation.created_at,
        "updated_at": escalation.updated_at,
        "resolved_at": escalation.resolved_at,
        "resolution_notes": escalation.resolution_notes,
    }




@router.get("/{escalation_id}")
def read_escalation(
    escalation_id: int,
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    escalation = get_escalation(
        db,
        current_user.hospital_id,
        escalation_id,
    )

    if not escalation:
        raise HTTPException(
            status_code=404,
            detail="Escalation not found",
        )

    evidence_dict = {}
    if escalation.evidence:
        try:
            evidence_dict = json.loads(escalation.evidence)
        except Exception:
            evidence_dict = {"raw": escalation.evidence}

    result = {
        "id": escalation.id,
        "hospital_id": escalation.hospital_id,
        "patient_id": escalation.patient_id,
        "queue_item_id": escalation.queue_item_id,
        "call_id": escalation.call_id,
        "consensus_assessment_id": (
            escalation.consensus_assessment_id
        ),
        "reason": escalation.reason,
        "priority": escalation.priority,
        "evidence": evidence_dict,
        "status": escalation.status,
        "source": escalation.source,
        "final_action": escalation.final_action,
        "created_at": escalation.created_at,
        "updated_at": escalation.updated_at,
        "resolved_at": escalation.resolved_at,
        "resolution_notes": escalation.resolution_notes,
    }

    return result


@router.get("/patient/{patient_id}")
def read_patient_escalations(
    patient_id: int,
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    escalations = get_patient_escalations(
        db,
        current_user.hospital_id,
        patient_id,
    )

    return [
        {
            "id": item.id,
            "patient_id": item.patient_id,
            "call_id": item.call_id,
            "consensus_assessment_id": (
                item.consensus_assessment_id
            ),
            "reason": item.reason,
            "priority": item.priority,
            "status": item.status,
            "final_action": item.final_action,
            "created_at": item.created_at,
            "resolved_at": item.resolved_at,
        }
        for item in escalations
    ]


@router.patch("/{escalation_id}/status")
def change_escalation_status(
    escalation_id: int,
    request: EscalationStatusUpdate,
    current_user=Depends(get_current_user),
    db: Session = Depends(get_db),
):
    escalation = get_escalation(
        db,
        current_user.hospital_id,
        escalation_id,
    )

    if not escalation:
        raise HTTPException(
            status_code=404,
            detail="Escalation not found",
        )

    try:
        escalation = update_escalation_status(
            db,
            escalation,
            request.status,
            request.resolution_notes,
        )

    except ValueError as exc:
        raise HTTPException(
            status_code=400,
            detail=str(exc),
        )

    evidence_dict = {}
    if escalation.evidence:
        try:
            evidence_dict = json.loads(escalation.evidence)
        except Exception:
            evidence_dict = {"raw": escalation.evidence}

    return {
        "id": escalation.id,
        "hospital_id": escalation.hospital_id,
        "patient_id": escalation.patient_id,
        "queue_item_id": escalation.queue_item_id,
        "call_id": escalation.call_id,
        "consensus_assessment_id": escalation.consensus_assessment_id,
        "reason": escalation.reason,
        "priority": escalation.priority,
        "evidence": json.dumps(evidence_dict) if isinstance(evidence_dict, dict) else escalation.evidence,
        "status": escalation.status,
        "source": escalation.source,
        "final_action": escalation.final_action,
        "created_at": escalation.created_at,
        "updated_at": escalation.updated_at,
        "resolved_at": escalation.resolved_at,
        "resolution_notes": escalation.resolution_notes,
    }

